shared/fetch-tc-artifacts.sh

```diff
@@ -26,12 +26,15 @@ REVISION="${REVISION_TREE}.revision.${INDEXED_HG_REV}"
 # a file and then feeding it to GNU parallel.
 echo "${CURL} https://index.taskcluster.net/v1/task/gecko.v2.$REVISION.source.source-bugzilla-info/artifacts/public/components-normalized.json > bugzilla-components.json" > downloads.lst
 for PLATFORM in linux64 macosx64 win64 android-armv7; do
-    # First check that the the searchfox job exists for the platform and revision we want. Otherwise emit a warning and skip it. This
-    # file is small so it's cheap to download and spew to stdout as a check that the analysis data for the platform exists.
-    ${CURL} https://index.taskcluster.net/v1/task/gecko.v2.$REVISION.firefox.$PLATFORM-searchfox-debug/artifacts/public/build/target.json ||
-    (   echo "WARNING: Unable to find analysis for $PLATFORM for hg rev $INDEXED_HG_REV; skipping analysis merge step for this platform." &&
-        continue
-    )
+    TC_PREFIX="https://index.taskcluster.net/v1/task/gecko.v2.${REVISION}.firefox.${PLATFORM}-searchfox-debug/artifacts/public/build"
+    # First check that the searchfox job exists for the platform and revision we want. Otherwise emit a warning and skip it. This
+    # file is small so it's cheap to download as a check that the analysis data for the platform exists.
+    #
+    # Also check for moz_source_stamp, to handle tasks that exists but failed. We rely on this field for resolve-gecko-revs.sh already.
+    if ! (${CURL} "${TC_PREFIX}/target.json" | grep moz_source_stamp); then
```

Comment:
Did you test the failure case here? Because of the `set -e` and `pipefail` options on the script if the grep command fails it will just abort the whole thing, I think.

Reply:
In Slack, Emilio said "Alright, so apparently that error check is just not working at all, as asking for garbage/target.json just returns a valid json file with a ResourceNotFound description".  So I think the idea is that as long as the network connection to taskcluster is good, any errors end up in the JSON rather than as HTTP status codes/etc.

Reply:
Yeah, I did (windows builds are busted in ash).

`pipefail` and `set -e` do account for if conditions. You can check this with:

```sh
#!/usr/bin/env bash

set -x # Show commands
set -eu # Errors/undefined vars are fatal
set -o pipefail # Check all commands in a pipeline

for foo in a b c; do
  if ! (echo "bar" | grep foo); then echo "Warning:";
    continue;
  fi
  echo $foo;
done
```

Interestingly, I learned that the `continue` as it was before cannot work at all, because parenthesis in bash effectively create a new environment (so the `continue` is "outside of the loop").

So this works:

```bash
for foo in a b c; do
  false || continue
  echo "should never reach this";
done
```

But this doesn't:

```bash
for foo in a b c; do
  false || (continue)
  echo "should never reach this";
done
```

So there were two reasons why this check wasn't working at all:

 * The json answer was always an http 200, which means that even if the task doesn't exist curl would never return a non-zero result.
 * The error handling code didn't work at all because of the parenthesis.

But anyhow yeah, I did test the error case :)

Reply:
Great, thanks for testing that. I've run into problems before with parentheses spawning subshells but didn't realize it was a problem here, thanks for catching that too :)